core/src/main/java/org/apache/calcite/plan/Convention.java

```diff
@@ -66,6 +67,13 @@ default boolean useAbstractConvertersForConversion(RelTraitSet fromTraits,
     return false;
   }
 
+  /**
+   * Return RelFactories struct for the convention which can be used to build RelNode
+   */
+  default RelFactories.Struct getRelFactories() {
+    return RelFactories.DEFAULT_STRUCT;
+  }
```

Comment:
Instead of add a `getRelFactories` interface to Convention, how about add a tool method `RelFactories#getRelFactories(Convention)`, it seems more straight forward because `RelFactories` is the factory to create all kinds of factory STRUCT.

Reply:
The thing is you don't want people to update RelFactories.java whenever there's a new convention available. Convention#getRelFactories seperates the logics, and can totally lives in user code base, instead of Calcite core.

Reply:
That's not a big problem, a builder to `RelFactories ` works.

Reply:
It seems complicate the design if we take that route, rather than "more straight forward".